File: src/recommendation/src/main/python/RankingTrainValidationSplit.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys

if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from mmlspark.Utils import *
from pyspark.ml.common import inherit_doc
from pyspark.ml import Model
from pyspark.ml.util import *
from multiprocessing.pool import ThreadPool

import numpy as np
import pyspark
import pyspark.sql.functions as F
import sys
from mmlspark.RankingSplitters import *
from mmlspark.RankingAdapter import *
from pyspark import keyword_only
from pyspark.ml import Estimator
from pyspark.ml.param import Params, Param, TypeConverters
from pyspark.ml.tuning import ValidatorParams
from pyspark.ml.util import *
from pyspark.sql import Window
from pyspark.sql.functions import col, expr
from pyspark.ml.param.shared import HasParallelism

if sys.version >= '3':
    basestring = str

def _parallelFitTasks(est, train, eva, validation, epm, collectSubModel):
    """
    Creates a list of callables which can be called from different threads to fit and evaluate
    an estimator in parallel. Each callable returns an `(index, metric)` pair.

    :param est: Estimator, the estimator to be fit.
    :param train: DataFrame, training data set, used for fitting.
    :param eva: Evaluator, used to compute `metric`
    :param validation: DataFrame, validation data set, used for evaluation.
    :param epm: Sequence of ParamMap, params maps to be used during fitting & evaluation.
    :param collectSubModel: Whether to collect sub model.
    :return: (int, float, subModel), an index into `epm` and the associated metric value.
    """
    modelIter = est.fitMultiple(train, epm)

    def singleTask():
        index, model = next(modelIter)
        metric = eva.evaluate(model.transform(validation, epm[index]))
        return index, metric, model if collectSubModel else None

    return [singleTask] * len(epm)


class HasCollectSubModels(Params):
    """
    Mixin for param collectSubModels: Param for whether to collect a list of sub-models trained during tuning. If set to false, then only the single best sub-model will be available after fitting. If set to true, then all sub-models will be available. Warning: For large models, collecting all sub-models can cause OOMs on the Spark driver.
    """

    collectSubModels = Param(Params._dummy(), "collectSubModels",
                             "Param for whether to collect a list of sub-models trained during tuning. If set to false, then only the single best sub-model will be available after fitting. If set to true, then all sub-models will be available. Warning: For large models, collecting all sub-models can cause OOMs on the Spark driver.",
                             typeConverter=TypeConverters.toBoolean)

    def __init__(self):
        super(HasCollectSubModels, self).__init__()
        self._setDefault(collectSubModels=False)

    def setCollectSubModels(self, value):
        """
        Sets the value of :py:attr:`collectSubModels`.
        """
        return self._set(collectSubModels=value)

    def getCollectSubModels(self):
        """
        Gets the value of collectSubModels or its default value.
        """
        return self.getOrDefault(self.collectSubModels)

@inherit_doc
class RankingTrainValidationSplit(Estimator, ValidatorParams, HasCollectSubModels, HasParallelism):
    trainRatio = Param(Params._dummy(), "trainRatio", "Param for ratio between train and\
         validation data. Must be between 0 and 1.", typeConverter=TypeConverters.toFloat)
    userCol = Param(Params._dummy(), "userCol",
                    "userCol: column name for user ids. Ids must be within the integer value range. (default: user)")
    ratingCol = Param(Params._dummy(), "ratingCol", "ratingCol: column name for ratings (default: rating)")

    itemCol = Param(Params._dummy(), "itemCol",
                    "itemCol: column name for item ids. Ids must be within the integer value range. (default: item)")

    def setTrainRatio(self, value):
        """
        Sets the value of :py:attr:`trainRatio`.
        """
        return self._set(trainRatio=value)

    def getTrainRatio(self):
        """
        Gets the value of trainRatio or its default value.
        """
        return self.getOrDefault(self.trainRatio)

    def setItemCol(self, value):
        """

        Args:

            itemCol (str): column name for item ids. Ids must be within the integer value range. (default: item)

        """
        self._set(itemCol=value)
        return self

    def getItemCol(self):
        """

        Returns:

            str: column name for item ids. Ids must be within the integer value range. (default: item)
        """
        return self.getOrDefault(self.itemCol)

    def setRatingCol(self, value):
        """

        Args:

            ratingCol (str): column name for ratings (default: rating)

        """
        self._set(ratingCol=value)
        return self

    def getRatingCol(self):
        """

        Returns:

            str: column name for ratings (default: rating)
        """
        return self.getOrDefault(self.ratingCol)

    def setUserCol(self, value):
        """

        Args:

            userCol (str): column name for user ids. Ids must be within the integer value range. (default: user)

        """
        self._set(userCol=value)
        return self

    def getUserCol(self):
        """

        Returns:

            str: column name for user ids. Ids must be within the integer value range. (default: user)
        """
        return self.getOrDefault(self.userCol)

    @keyword_only
    def __init__(self, estimator=None, estimatorParamMaps=None, evaluator=None, seed=None, trainRatio=0.8):
        """
        __init__(self, estimator=None, estimatorParamMaps=None, evaluator=None, numFolds=3,\
                 seed=None)
        """
        super(RankingTrainValidationSplit, self).__init__()
        kwargs = self._input_kwargs
        self._set(**kwargs)

    @keyword_only
    def setParams(self, estimator=None, estimatorParamMaps=None, evaluator=None, seed=None):
        """
        setParams(self, estimator=None, estimatorParamMaps=None, evaluator=None, numFolds=3,\
                  seed=None):
        Sets params for cross validator.
        """
        kwargs = self._input_kwargs
        return self._set(**kwargs)

    def copy(self, extra=None):
        """
        Creates a copy of this instance with a randomly generated uid
        and some extra params. This copies creates a deep copy of
        the embedded paramMap, and copies the embedded and extra parameters over.

        :param extra: Extra parameters to copy to the new instance
        :return: Copy of this instance
        """
        if extra is None:
            extra = dict()
        newCV = Params.copy(self, extra)
        if self.isSet(self.estimator):
            newCV.setEstimator(self.getEstimator().copy(extra))
        # estimatorParamMaps remain the same
        if self.isSet(self.evaluator):
            newCV.setEvaluator(self.getEvaluator().copy(extra))
        return newCV

    def _create_model(self, java_model):
        model = RankingTrainValidationSplitModel()
        model._java_obj = java_model
        model._transfer_params_from_java()
        return model

    def split(self, dataset, tRatio, userColumn, itemColumn):
        pyspark.sql.DataFrame.min_rating_filter = RankingSplitters.min_rating_filter
        pyspark.sql.DataFrame.stratified_split = RankingSplitters.stratified_split

        temp_train, temp_validation = dataset \
            .dropDuplicates() \
            .withColumnRenamed(userColumn, 'customerID') \
            .withColumnRenamed(itemColumn, 'itemID') \
            .min_rating_filter(min_rating=6, by_customer=True) \
            .stratified_split(min_rating=3, by_customer=True, fixed_test_sample=False, ratio=tRatio)

        train = temp_train \
            .withColumnRenamed('customerID', userColumn) \
            .withColumnRenamed('itemID', itemColumn)

        validation = temp_validation \
            .withColumnRenamed('customerID', userColumn) \
            .withColumnRenamed('itemID', itemColumn)
        return train, validation


    def _fit(self, dataset):

        rating = self.getOrDefault(self.estimator).getRatingCol()
        userColumn = self.getOrDefault(self.estimator).getUserCol()
        itemColumn = self.getOrDefault(self.estimator).getItemCol()

        eva = self.getOrDefault(self.evaluator)

        est = RankingAdapter() \
            .setRecommender(self.getOrDefault(self.estimator)) \
            .setMode("allUsers") \
            .setK(eva) \
            .setUserCol(userColumn) \
            .setItemCol(itemColumn) \
            .setRatingCol(rating)

        epm = self.getOrDefault(self.estimatorParamMaps)
        numModels = len(epm)

        tRatio = self.getOrDefault(self.trainRatio)
        seed = self.getOrDefault(self.seed)

        train, validation = self.split(dataset, tRatio, userColumn, itemColumn)

        subModels = None
        collectSubModelsParam = self.getCollectSubModels()
        if collectSubModelsParam:
            subModels = [None for i in range(numModels)]

        tasks = _parallelFitTasks(est, train, eva, validation, epm, collectSubModelsParam)
        pool = ThreadPool(processes=min(self.getParallelism(), numModels))
        metrics = [None] * numModels
        for j, metric, subModel in pool.imap_unordered(lambda f: f(), tasks):
            metrics[j] = metric
            if collectSubModelsParam:
                subModels[j] = subModel

        train.unpersist()
        validation.unpersist()

        if eva.isLargerBetter():
            bestIndex = np.argmax(metrics)
        else:
            bestIndex = np.argmin(metrics)
        bestModel = est.fit(dataset, epm[bestIndex])
        return self._copyValues(RankingTrainValidationSplitModel(bestModel, metrics, subModels))

    # @classmethod
    # def _from_java(cls, java_stage):
    #     """
    #     Given a Java TrainValidationSplit, create and return a Python wrapper of it.
    #     Used for ML persistence.
    #     """
    #
    #     estimator, epms, evaluator = super(TrainValidationSplit, cls)._from_java_impl(java_stage)
    #     trainRatio = java_stage.getTrainRatio()
    #     seed = java_stage.getSeed()
    #     parallelism = java_stage.getParallelism()
    #     collectSubModels = java_stage.getCollectSubModels()
    #     # Create a new instance of this stage.
    #     py_stage = cls(estimator=estimator, estimatorParamMaps=epms, evaluator=evaluator,
    #                    trainRatio=trainRatio, seed=seed, parallelism=parallelism,
    #                    collectSubModels=collectSubModels)
    #     py_stage._resetUid(java_stage.uid())
    #     return py_stage
    #
    # def _to_java(self):
    #     """
    #     Transfer this instance to a Java TrainValidationSplit. Used for ML persistence.
    #     :return: Java object equivalent to this instance.
    #     """
    #
    #     estimator, epms, evaluator = super(TrainValidationSplit, self)._to_java_impl()
    #
    #     _java_obj = JavaParams._new_java_obj("org.apache.spark.ml.tuning.TrainValidationSplit",
    #                                          self.uid)
    #     _java_obj.setEstimatorParamMaps(epms)
    #     _java_obj.setEvaluator(evaluator)
    #     _java_obj.setEstimator(estimator)
    #     _java_obj.setTrainRatio(self.getTrainRatio())
    #     _java_obj.setSeed(self.getSeed())
    #     _java_obj.setParallelism(self.getParallelism())
    #     _java_obj.setCollectSubModels(self.getCollectSubModels())
    #     return _java_obj


@inherit_doc
class RankingTrainValidationSplitModel(Model, ValidatorParams):

    def __init__(self, bestModel, validationMetrics=[], subModels=None):
        super(RankingTrainValidationSplitModel, self).__init__()
        #: best model from cross validation
        self.bestModel = bestModel
        #: evaluated validation metrics
        self.validationMetrics = validationMetrics
        self.subModels = subModels

    def _transform(self, dataset):
        return self.bestModel.transform(dataset)

    def copy(self, extra=None):
        """
        Creates a copy of this instance with a randomly generated uid
        and some extra params. This copies the underlying bestModel,
        creates a deep copy of the embedded paramMap, and
        copies the embedded and extra parameters over.
        And, this creates a shallow copy of the validationMetrics.

        :param extra: Extra parameters to copy to the new instance
        :return: Copy of this instance
        """
        if extra is None:
            extra = dict()
        bestModel = self.bestModel.copy(extra)
        validationMetrics = list(self.validationMetrics)
        return RankingTrainValidationSplitModel(bestModel, validationMetrics)

    def recommendForAllUsers(self, numItems):
        return self.bestModel.recommendForAllUsers(numItems)
        # return self.bestModel._call_java("recommendForAllUsers", numItems)

    def recommendForAllItems(self, numItems):
        return self.bestModel.recommendForAllItems(numItems)

    # @classmethod
    # def _from_java(cls, java_stage):
    #     """
    #     Given a Java TrainValidationSplitModel, create and return a Python wrapper of it.
    #     Used for ML persistence.
    #     """
    #
    #     # Load information from java_stage to the instance.
    #     bestModel = JavaParams._from_java(java_stage.bestModel())
    #     estimator, epms, evaluator = super(TrainValidationSplitModel,
    #                                        cls)._from_java_impl(java_stage)
    #     # Create a new instance of this stage.
    #     py_stage = cls(bestModel=bestModel).setEstimator(estimator)
    #     py_stage = py_stage.setEstimatorParamMaps(epms).setEvaluator(evaluator)
    #
    #     if java_stage.hasSubModels():
    #         py_stage.subModels = [JavaParams._from_java(sub_model)
    #                               for sub_model in java_stage.subModels()]
    #
    #     py_stage._resetUid(java_stage.uid())
    #     return py_stage
    #
    # def _to_java(self):
    #     """
    #     Transfer this instance to a Java TrainValidationSplitModel. Used for ML persistence.
    #     :return: Java object equivalent to this instance.
    #     """
    #
    #     sc = SparkContext._active_spark_context
    #     # TODO: persst validation metrics as well
    #     _java_obj = JavaParams._new_java_obj(
    #         "org.apache.spark.ml.tuning.TrainValidationSplitModel",
    #         self.uid,
    #         self.bestModel._to_java(),
    #         _py2java(sc, []))
    #     estimator, epms, evaluator = super(TrainValidationSplitModel, self)._to_java_impl()
    #
    #     _java_obj.set("evaluator", evaluator)
    #     _java_obj.set("estimator", estimator)
    #     _java_obj.set("estimatorParamMaps", epms)
    #
    #     if self.subModels is not None:
    #         java_sub_models = [sub_model._to_java() for sub_model in self.subModels]
    #         _java_obj.setSubModels(java_sub_models)
    #
    #     return _java_obj

File: src/recommendation/src/main/scala/RankingTrainValidationSplit.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark

import org.apache.spark.ml._
import org.apache.spark.ml.evaluation.Evaluator
import org.apache.spark.ml.param.{IntParam, ParamMap, ParamValidators}
import org.apache.spark.ml.recommendation._
import org.apache.spark.ml.util._
import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.{DataFrame, Dataset, Row}

import scala.collection.mutable
import scala.concurrent.duration.Duration
import scala.concurrent.{ExecutionContext, Future}
import scala.util.Random
import org.apache.spark.ml.evaluation.Evaluator
import org.apache.spark.ml.param._
import org.apache.spark.ml.param.shared.HasCollectSubModels
import org.apache.spark.ml.recommendation.ALS
import org.apache.spark.ml.tuning.{RankingHelper, TrainValidRecommendSplitParams}
import org.apache.spark.ml.util._
import org.apache.spark.ml.{Estimator, Model, Transformer}
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.util.ThreadUtils

import scala.concurrent.Future
import scala.concurrent.duration.Duration
import scala.language.existentials

/**
  * Validation for hyper-parameter tuning.
  * Randomly splits the input dataset into train and validation sets,
  * and uses evaluation metric on the validation set to select the best model.
  */
@InternalWrapper
class RankingTrainValidationSplit(override val uid: String)
  extends Estimator[RankingTrainValidationSplitModel]
    with ComplexParamsWritable with RankingFunctions
    with TrainValidRecommendSplitParams with HasCollectSubModels {

  /** @group setParam */
  def setEstimator(value: Estimator[_ <: Model[_]]): this.type = set(estimator, value)

  /** @group setParam */
  def setEstimatorParamMaps(value: Array[ParamMap]): this.type = set(estimatorParamMaps, value)

  /** @group setParam */
  def setEvaluator(value: Evaluator): this.type = set(evaluator, value)

  /** @group setParam */
  def setTrainRatio(value: Double): this.type = set(trainRatio, value)

  /** @group setParam */
  def setSeed(value: Long): this.type = set(seed, value)

  /**
    * Set the mamixum level of parallelism to evaluate models in parallel.
    * Default is 1 for serial evaluation
    *
    * @group expertSetParam
    */
  def setParallelism(value: Int): this.type = set(parallelism, value)

  /**
    * Whether to collect submodels when fitting. If set, we can get submodels from
    * the returned model.
    *
    * Note: If set this param, when you save the returned model, you can set an option
    * "persistSubModels" to be "true" before saving, in order to save these submodels.
    *
    * @group expertSetParam
    */
  def setCollectSubModels(value: Boolean): this.type = set(collectSubModels, value)

  val collectSubMetrics: BooleanParam = new BooleanParam(this, "collectSubMetrics", "")

  def setCollectSubMetrics(value: Boolean): this.type = set(collectSubMetrics, value)

  def getCollectSubMetrics: Boolean = $(collectSubMetrics)

  def this() = this(Identifiable.randomUID("RankingTrainValidationSplit"))

  def transformSchema(schema: StructType): StructType = {
    require($(estimatorParamMaps).nonEmpty, s"Validator requires non-empty estimatorParamMaps")
    val firstEstimatorParamMap = $(estimatorParamMaps).head
    val est = $(estimator)
    for (paramMap <- $(estimatorParamMaps).tail) {
      est.copy(paramMap).transformSchema(schema)
    }
    est.copy(firstEstimatorParamMap).transformSchema(schema)
  }

  setDefault(minRatingsPerUser -> 1, minRatingsPerItem -> 1)

  override def getExecutionContext: ExecutionContext = {

    getParallelism match {
      case 1 =>
        RankingHelper.getThreadUtils().sameThread
      case n =>
        ExecutionContext.fromExecutorService(RankingHelper.getThreadUtils()
          .newDaemonCachedThreadPool(s"${this.getClass.getSimpleName}-thread-pool", n))
    }
  }

  def fit(dataset: Dataset[_]): RankingTrainValidationSplitModel = {
    val schema = dataset.schema
    transformSchema(schema)

    val eval = $(evaluator).asInstanceOf[RankingEvaluator]

    val est = new RankingAdapter()
      .setMode("allUsers") //allItems does not work, not sure if it would be used
      .setK(eval.getK)
      .setRecommender($(estimator).asInstanceOf[Estimator[_ <: Model[_]]])
      .setUserCol($(estimator).asInstanceOf[ALS].getUserCol)
      .setRatingCol($(estimator).asInstanceOf[ALS].getRatingCol)
      .setItemCol($(estimator).asInstanceOf[ALS].getItemCol)
    //todo cast to something more generic than ALS

    val epm = $(estimatorParamMaps)

    // Create execution context based on $(parallelism)
    val executionContext = getExecutionContext

    val (trainingDataset, validationDataset) =
      split(dataset, getTrainRatio, est.getItemCol, est.getUserCol, est.getRatingCol)
    trainingDataset.cache()
    validationDataset.cache()

    val collectSubModelsParam = $(collectSubModels)
    val collectSubMetricsParam = $(collectSubMetrics)

    val subModels: Option[Array[Model[_]]] = if (collectSubModelsParam) {
      Some(Array.fill[Model[_]](epm.length)(null))
    } else None

    val subMetrics: Option[Array[Map[String, Double]]] = if (collectSubMetricsParam) {
      Some(Array.fill[Map[String, Double]](epm.length)(null))
    } else None

    // Fit models in a Future for training in parallel
    logDebug(s"Train split with multiple sets of parameters.")
    val metricFutures = epm.zipWithIndex.map { case (paramMap, paramIndex) =>
      Future[Double] {
        val model = est.fit(trainingDataset, paramMap).asInstanceOf[Model[_]]

        if (collectSubModelsParam) {
          subModels.get(paramIndex) = model
        }
        // TODO: duplicate evaluator to take extra params from input
        val df = model.transform(validationDataset, paramMap)
        if (collectSubMetricsParam) {
          subMetrics.get(paramIndex) = eval.asInstanceOf[RankingEvaluator].getMetricsMap(df)
        }
        val metric = eval.evaluate(df)
        logDebug(s"Got metric $metric for model trained with $paramMap.")
        metric
      }(executionContext)
    }

    // Wait for all metrics to be calculated
    val metrics = metricFutures.map(RankingHelper.getThreadUtils().awaitResult(_, Duration.Inf))

    // Unpersist training & validation set once all metrics have been produced
    trainingDataset.unpersist()
    validationDataset.unpersist()

    logInfo(s"Train validation split metrics: ${metrics.toSeq}")
    val (bestMetric, bestIndex) =
      if (eval.isLargerBetter) metrics.zipWithIndex.maxBy(_._1)
      else metrics.zipWithIndex.minBy(_._1)
    logInfo(s"Best set of parameters:\n${epm(bestIndex)}")
    logInfo(s"Best train validation split metric: $bestMetric.")
    val bestModel = est.fit(dataset, epm(bestIndex)).asInstanceOf[Model[_]]
    new RankingTrainValidationSplitModel(uid)
      .setBestModel(bestModel)
      .setMetrics(metrics)
      .setSubModels(subModels)
      .setSubMetrics(subMetrics)
  }

  override def copy(extra: ParamMap): RankingTrainValidationSplit = {
    defaultCopy(extra)
  }

}

object RankingTrainValidationSplit extends ComplexParamsReadable[RankingTrainValidationSplit]

/**
  * Model from train validation split.
  *
  * @param uid Id.
  */
@InternalWrapper
class RankingTrainValidationSplitModel private[spark](val uid: String)
  extends Model[RankingTrainValidationSplitModel]
    with ComplexParamsWritable with Wrappable {

  private var _subModels: Option[Array[Model[_]]] = None

  private[spark] def setSubModels(subModels: Option[Array[Model[_]]])
  : RankingTrainValidationSplitModel = {
    _subModels = subModels
    this
  }

  /**
    * @return submodels represented in array. The index of array corresponds to the ordering of
    *         estimatorParamMaps
    */
  def subModels: Array[Model[_]] = {
    require(_subModels.isDefined, "subModels not available, To retrieve subModels, make sure " +
      "to set collectSubModels to true before fitting.")
    _subModels.get
  }

  def hasSubModels: Boolean = _subModels.isDefined

  private var _subMetrics: Option[Array[Map[String, Double]]] = None

  private[spark] def setSubMetrics(subMetrics: Option[Array[Map[String, Double]]])
  : RankingTrainValidationSplitModel = {
    _subMetrics = subMetrics
    this
  }

  /**
    * @return submodels represented in array. The index of array corresponds to the ordering of
    *         estimatorParamMaps
    */
  def subMetrics: Array[Map[String, Double]] = {
    require(_subMetrics.isDefined, "subModels not available, To retrieve subModels, make sure " +
      "to set collectSubModels to true before fitting.")
    _subMetrics.get
  }

  def hasSubMetrics: Boolean = _subMetrics.isDefined

  def recommendForAllUsers(k: Int): DataFrame = getBestModel.asInstanceOf[RankingAdapterModel]
    .recommendForAllUsers(k)

  def recommendForAllItems(k: Int): DataFrame = getBestModel.asInstanceOf[RankingAdapterModel]
    .recommendForAllItems(k)

  val metrics = new DoubleArrayParam(this, "metrics", "metrics")

  def setMetrics(m: Array[Double]): RankingTrainValidationSplitModel.this.type = set(metrics, m)

  def getMetrics: Array[Double] = $(metrics)

  val bestModel = new TransformerParam(this, "bestModel", "bestModel", { x: Transformer => true })

  def setBestModel(m: Model[_]): RankingTrainValidationSplitModel.this.type = set(bestModel, m)

  def getBestModel: Model[_] = $(bestModel).asInstanceOf[Model[_]]

  def transform(dataset: Dataset[_]): DataFrame = {
    transformSchema(dataset.schema)
    getBestModel.transform(dataset)
  }

  def transformSchema(schema: StructType): StructType = {
    getBestModel.transformSchema(schema)
  }

  override def copy(extra: ParamMap): RankingTrainValidationSplitModel = {
    defaultCopy(extra)
  }

}

object RankingTrainValidationSplitModel extends ComplexParamsReadable[RankingTrainValidationSplitModel]